lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,959 @@
+  // the module is not instantiated or is multiply instantiated.
+  auto exactlyOneInstance = [&](FModuleOp op,
+                                StringRef msg) -> Optional<InstanceOp> {
+    auto instances = symbolTable.getSymbolUses(op, circuitOp);
+    if (!instances || std::distance(instances.getValue().begin(),
+                                    instances.getValue().end()) == 0) {
+      op->emitOpError() << "is marked as a GrandCentral '" << msg
+                        << "', "
+                           "but is never instantiated";
+      return None;
     }
+    if (std::distance(instances.getValue().begin(),
+                      instances.getValue().end()) != 1) {
+      auto diag = op->emitOpError() << "is marked as a GrandCentral '" << msg
+                                    << "', but "
+                                       "it is instantiated more than once";
+      for (auto instance : instances.getValue())
+        diag.attachNote(instance.getUser()->getLoc())
+            << "parent is instantiated here";
+    }
+    return cast<InstanceOp>((*(instances.getValue().begin())).getUser());
+  };
 
-    // If the interface already exists, don't create it.
-    if (interfaces.count(bundle.defName))
-      return true;
-
-    // Create the interface.  This will be populated later.
-    interfaces[bundle.defName] =
-        builder.create<sv::InterfaceOp>(circuitOp->getLoc(), bundle.defName);
-
-    return true;
+  /// Walk the circuit and extract all information related to scattered
+  /// Grand Central annotations.  This is used to populate: (1) the
+  /// companionIDMap, (2) the parentIDMap, and (3) the leafMap.
+  /// Annotations are removed as they are discovered and if they are not
+  /// malformed.
+  removalError = false;
+  circuitOp.walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op)
+        .Case<RegOp, RegResetOp, WireOp, NodeOp>([&](auto op) {
+          AnnotationSet::removeAnnotations(op, [&](Annotation annotation) {
+            if (!annotation.isClass(
+                    "sifive.enterprise.grandcentral.AugmentedGroundType"))
+              return false;
+            auto maybeID = getID(op, annotation);
+            if (!maybeID)
+              return false;
+            leafMap.insert({maybeID.getValue(), op.getResult()});
+            return true;
+          });
+        })
+        // TODO: Figure out what to do with this.
+        .Case<InstanceOp>([&](auto op) {
+          AnnotationSet::removePortAnnotations(op, [&](unsigned i,
+                                                       Annotation annotation) {
+            if (!annotation.isClass(
+                    "sifive.enterprise.grandcentral.AugmentedGroundType"))
+              return false;
+            op.emitOpError()
+                << "is marked as an interface element, but this should be "
+                   "impossible due to how the Chisel Grand Central API works";
```

Comment:
I have *no* idea how anyone else could have written this pass, without all that GCT insight. Holy cow.

Reply:
😅 I only came across this via experimentation... There are two weird possibilities here that I'm just punting on: 

1) (The one mentioned here...) The Grand Central Chisel API relies on a typeclass implementation that defines how the type of the thing in the interface is converted to the interface. If that doesn't exist it bails out. Trying to add a `mem` to the interface will then fail with something like:

```